### ISSUE TITLE
content.json distribution problem - 1 second dif - optional files issue?

### ISSUE THREAD

**Issue by jaros1** [HelloZeroNet/ZeroNet#1173]
I am trying to use optional files in ZeroNet. 
Testing on two different machines with vpn=torando and vpn=sweden and zeronet port open. 
I would expect newest content.json file to be distributed to other clients on ZeroNet.

I have many times seen the situation where there is a very small difference in content.json modified timestamp (1 second) and newest content.json is not distributed on zeronet. I have seen this issue before and the 1 second difference between the two content.json files cannot be something random. Must be UI server doing something "funny". 

Maybe a "merge" problem due to different files_optional in the two content.json?

Torando content.json (newest):

<pre>
{
 "address": "1HXzvtSLuvxZfh6LgdaqTk4FSVf7x8w7NJ",
 "cert_auth_type": "web",
 "cert_sign": "G9KWZvNAg3yVm3F3ICPQCOjnA+lZQOFLvYWBJp7BaRAMDrdIAqyv/x8PYNwlWQ4KvznJPOmjknSxxPkHfnD84tE=",
 "cert_user_id": "jro@zeroid.bit",
 "files": {
  "status.json": {
   "sha512": "ea18a2cce60357f5a647eef13bd0701a418f63795083226dbabd7f62e32a8e48",
   "size": 31
  },
  "wallet.json": {
   "sha512": "5033a27b0aba76b024e04f037eff88a121781dd33d292fc229734f0b9dff6c64",
   "size": 869
  }
 },
 "files_optional": {
  "7fd09cfaf8-o.1509785536712": {
   "sha512": "9a53ceadb1336c504692dbd223ba1d2c089d2018122a00e5146a6110a87a8974",
   "size": 548
  },
  "7fd09cfaf8-o.1509785537477": {
   "sha512": "765f493bc022b38138817e508ca5301d31d603273604748c07d1a2547b0a3e22",
   "size": 1424
  }
 },
 "inner_path": "data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json",
 "modified": 1509785537,
 "optional": "^.*-.*$",
 "signs": {
  "18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ": "HCEb8ZkEG29y8BHxK4sjascvZuk/Z20tl4pz++r5j8bJTqog0O2KRqCnJtG65eZ+UkR8MCY/aE6PSlSbz1wwVAk="
 }
}
</pre>

UI server log sweden: 

torando has just published 09:52:17 new data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json with new optional file 7fd09cfaf8-o.1509785537477

<pre>
[09:52:16] WorkerManager:1PgyTn..z2zh New task: data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json, peer lock: None, priority: 2, optional_hash_id: None, tasks started: 1
[09:52:17] Site:1HXzvt..w7NJ Update for data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json looks valid, saving...
[09:52:17] Site:1HXzvt..w7NJ Patched successfully: data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/status.json (diff: 0.000s, verify: 0.000s, write: 0.001s, on_done: 0.001s)
[09:52:17] Site:1HXzvt..w7NJ Publishing data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json to 3/15 peers (connected: 4) diffs: ['status.json'] (0.09k)...
[09:52:17] Site:1HXzvt..w7NJ Peer:173.239.230.28 already received this update for data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json, skipping
[09:52:21] Site:1PgyTn..z2zh Bad files: {'data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json': 2}
[09:52:21] Site:1PgyTn..z2zh Retry 1 bad files
[09:52:28] Site:1HXzvt..w7NJ Successfuly data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json published to 2 peers, publishing to 3 more peers in the background
[09:52:40] WorkerManager:1PgyTn..z2zh 37.203.209.19:39492: Verify failed: data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json, error: Download failed, failed peers: 0, workers: 5
[09:52:40] WorkerManager:1PgyTn..z2zh Slow task: data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json, (workers: 6, optional_hash_id: None, peers: 0, failed: 1, asked: 0)
[09:52:47] Site:1HXzvt..w7NJ data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json: verify sign error: VerifyError: Valid signs: 0/1 in ContentManager.py line 861
[09:52:47] FileServer Update for data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json is invalid: Valid signs: 0/1
[09:52:47] RateLimit Added to queue (14.69s left): 16 update 1HXzvtSLuvxZfh6LgdaqTk4FSVf7x8w7NJ data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json 
[09:52:56] WorkerManager:1PgyTn..z2zh Slow task: data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json, (workers: 5, optional_hash_id: None, peers: 0, failed: 1, asked: 0)
[09:52:56] Site:1HXzvt..w7NJ Announced types ['ip4'] in mode more to 1 trackers in 0.349s, errors: [], slow: []
[09:53:02] RateLimit Calling: 16 update 1HXzvtSLuvxZfh6LgdaqTk4FSVf7x8w7NJ data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json
[09:53:02] Site:1HXzvt..w7NJ data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json: verify sign error: VerifyError: Valid signs: 0/1 in ContentManager.py line 861
[09:53:02] FileServer Update for data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json is invalid: Valid signs: 0/1
[09:53:10] WorkerManager:1HXzvt..w7NJ 77.241.139.79:37344: Verify failed: data/users/1Mg62SVtQAiRSAfRBvWGKyPqioXZSfwDDW/wallet.json, error: Download failed, failed peers: 9, workers: 1
[09:53:10] WorkerManager:1PgyTn..z2zh 212.27.18.166:28073: Verify failed: data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json, error: Download failed, failed peers: 1, workers: 4
[09:53:10] WorkerManager:1PgyTn..z2zh 183.27.96.134:15441: Verify failed: data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json, error: Download failed, failed peers: 2, workers: 3
[09:53:10] WorkerManager:1PgyTn..z2zh 183.27.96.243:15441: Verify failed: data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json, error: Download failed, failed peers: 3, workers: 2
[09:53:10] WorkerManager:1PgyTn..z2zh gbr6xj5fdxczy6jr.onion:39492: Verify failed: data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json, error: Download failed, failed peers: 4, workers: 1
[09:53:10] WorkerManager:1PgyTn..z2zh t27vym4lfo3xjuvc.onion:39492: Verify failed: data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json, error: Download failed, failed peers: 5, workers: 0
[09:53:10] WorkerManager:1HXzvt..w7NJ ljvtyp4ipykefote.onion:39492: Verify failed: data/users/1Mg62SVtQAiRSAfRBvWGKyPqioXZSfwDDW/wallet.json, error: Download failed, failed peers: 10, workers: 0
[09:53:11] WorkerManager:1PgyTn..z2zh Slow task: data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json, (workers: 5, optional_hash_id: None, peers: 0, failed: 6, asked: 0)
[09:53:12] Site:1PgyTn..z2zh data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json: verify sign error: VerifyError: Valid signs: 0/1 in ContentManager.py line 861
[09:53:12] WorkerManager:1PgyTn..z2zh 173.239.230.28:45260: Verify failed: data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json, error: Valid signs: 0/1, failed peers: 6, workers: 0
[09:53:12] WorkerManager:1PgyTn..z2zh 127.0.0.1:60838: Verify failed: data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json, error: Download failed, failed peers: 7, workers: 4
[09:53:12] WorkerManager:1PgyTn..z2zh 104.199.246.69:15441: Verify correct: data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json
[09:53:12] Site:1PgyTn..z2zh data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json loadContent same json file, skipping
[09:53:12] Site:1PgyTn..z2zh data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json loadContent same json file, skipping
</pre>

Sweden content.json (oldest):

<pre>
{
 "address": "1HXzvtSLuvxZfh6LgdaqTk4FSVf7x8w7NJ",
 "cert_auth_type": "web",
 "cert_sign": "G9KWZvNAg3yVm3F3ICPQCOjnA+lZQOFLvYWBJp7BaRAMDrdIAqyv/x8PYNwlWQ4KvznJPOmjknSxxPkHfnD84tE=",
 "cert_user_id": "jro@zeroid.bit",
 "files": {
  "status.json": {
   "sha512": "a84a76dfca3849cf7b930852c724cd0f5ab58bb173e64b4f666abc2bfc889b33",
   "size": 31
  },
  "wallet.json": {
   "sha512": "5033a27b0aba76b024e04f037eff88a121781dd33d292fc229734f0b9dff6c64",
   "size": 869
  }
 },
 "files_optional": {
  "7fd09cfaf8-o.1509785536712": {
   "sha512": "9a53ceadb1336c504692dbd223ba1d2c089d2018122a00e5146a6110a87a8974",
   "size": 548
  },
  "c6f8666c5b-i.1477976389761": {
   "sha512": "5ce44bfd2791b58cea417692c10cedc22698cfbe6615e29bf8ca04735fe7e300",
   "size": 1316
  },
  "c6f8666c5b-i.1479193655354": {
   "sha512": "7a5b61ce3aa612fbe6fcdaca85d6e0f4524144c60e3857324560890f426e9142",
   "size": 1316
  },
  "c6f8666c5b-i.1480073346388": {
   "sha512": "08993fa11262ab18c88badb97605f6d70386bf67ac70ec13f1051f66b658d3e9",
   "size": 1316
  }
 },
 "inner_path": "data/users/18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ/content.json",
 "modified": 1509785536,
 "optional": "^.*-.*$",
 "signs": {
  "18DbeZgtVCcLghmtzvg4Uv8uRQAwR8wnDQ": "G7etvqxwJRx+jy7v2QG+GgZ0KklUIGfJM7S4nBPcs1BLCQ0Emm/ANc99bQADsr2fmRy+q4RCuY31LtXIh7X7Kuw="
 }
}
</pre>

compare content.json files:
![screenshot from 2017-11-04 10-23-30](https://user-images.githubusercontent.com/4932088/32404566-0a2484aa-c152-11e7-96da-0d9b0c21ce9b.png)

**Comment by HelloZeroNet** [comment]
There is a rate-limiting in client: to avoid network attacks a client won't accept modification on the same file within 15sec.
Normally your client should automatically queue and delay updates.

**Comment by jaros1** [comment]
Thank you. The 15 sec limit can explain my problems.

Regarding 1HXzvtSLuvxZfh6LgdaqTk4FSVf7x8w7NJ and 1PgyTnnACGd1XRdpfiDihgKwYRRnzgz2zh. 
They are both MoneyNetwork merger sites.

I have one running site MoneyNetwork using 1PgyTnnACGd1XRdpfiDihgKwYRRnzgz2zh as a hub 
and an other running site MoneyNetworkW2 using 1HXzvtSLuvxZfh6LgdaqTk4FSVf7x8w7NJ as a hub.
The sign/publish operations in the two sites should use different user directories but I have to check my logs. But the two sites may sign and publish content to fast to be accepted by other clients. I will add some extra js code to keep publish interval > 15 seconds.

**Comment by jaros1** [comment]
I know that ZeroFrame JS code has a 30 seconds publish queue. Very nice UI and okay when publishing from one site but this does not work when publishing from 2 sites at the same time in two different windows or browser tabs. 

Is it possible to let the UI server return an ratelimit error when publishing content too fast? 
Now the UI server returns OK publish to the client but the published content.json file will be rejected by other ZeroNet clients due to ratelimit. So the OK published is really a failed publish.

**Comment by HelloZeroNet** [comment]
It's global per-client and does not depends on browser window.

**Comment by jaros1** [comment]
Yes. I have for now MoneyNetwork and MoneyNetworkW2 that is merger sites using hubs with merger_type=MoneyNetwork. Publish in both sites at the same time. 

I am working on a JS queue to prevent getting below the ratelimit when publishing from two sites but a "little" hard to implement. Do not yet know if my special JS queue will work. Maybe easier to catch in the UI server and return Publish failed: ratelimit or something like this to the UI?

I am sure that this is a very rare problem but I have spend many hours before learning that ratelimit while publishing was the problem and that publish OK not already is publish OK.

**Comment by jaros1** [comment]
Regarding ratelimit check. I guess that I can use content.modified timestamp + 15 seconds before starting next publish?

**Comment by jaros1** [comment]
Added a publish queue in MoneyNetwork with 15 seconds between publish and content.json files are being distributed OK now. Closing this issue but I still  think that the error (ratelimit) should be reported already at publish time and not just being blocked be receiver peers

**Comment by jaros1** [comment]
Run into the same problem with 26 seconds dif between the two content.json files. Have increased my internal publish queue wait to 30 seconds and everything looks fine now. Thank you for your help.

**Comment by HelloZeroNet** [comment]
It's recommended to keep it longer, because the update does not reaches everyone immediately.
